File: apps/ios/NARTHECare/Models/HealthObservation.swift
import Foundation

/// Swift mirror of `shared/contracts/healthObservation.schema.json`.
///
/// This file is the iOS side of the cross-platform sync contract for
/// `POST /healthkit/sync`. It is hand-kept in lockstep with:
///
/// - `shared/contracts/healthObservation.schema.json` (source of truth)
/// - `shared/models/HealthObservation.ts` (web TS mirror)
/// - `shared/models/HealthObservation.js` (backend JS mirror)
///
/// Field names are camelCase and match the JSON payload exactly so no
/// `CodingKeys` remapping is needed — same convention as
/// `CareRecipientProfile.swift`. Enum raw values are snake_case to
/// match the schema's enum strings.
///
/// All structs are `Codable, Sendable` so SwiftUI views and view models
/// can hand them across `Task`/actor boundaries without warnings.
/// **Never log instances of these types** — `value`, `measuredAt`,
/// `metadata`, and `sourceRecordId` are PHI under the HIPAA framing
/// this app operates under.

// MARK: - Enums

/// Transport that delivered the sample. Phase 4A only ships
/// `.healthkit` from iOS; `.manual` is reserved for future
/// caregiver-entered rows.
enum HealthObservationSourceType: String, Codable, Sendable, CaseIterable {
  case healthkit
  case manual
}

/// Canonical metric_type strings persisted in
/// `health_observations.metric_type` on the backend.
enum HealthObservationMetricType: String, Codable, Sendable, CaseIterable {
  case steps
  case restingHeartRate = "resting_heart_rate"
  case hrv
  case spo2
  case sleepDuration = "sleep_duration"
  case respiratoryRate = "respiratory_rate"
  case walkingSteadiness = "walking_steadiness"
  case fallEvent = "fall_event"

  /// The unit string that must accompany this metric in the contract.
  /// Encoded once here so the sync service cannot accidentally pair
  /// a metric with the wrong unit (the backend rejects mismatches
  /// with a 400).
  var unit: HealthObservationUnit {
    switch self {
    case .steps: return .count
    case .restingHeartRate: return .bpm
    case .hrv: return .ms
    case .spo2: return .percent
    case .sleepDuration: return .hours
    case .respiratoryRate: return .breathsPerMin
    case .walkingSteadiness: return .score
    case .fallEvent: return .event
    }
  }
}

/// Canonical units paired with each metric type.
enum HealthObservationUnit: String, Codable, Sendable, CaseIterable {
  case count
  case bpm
  case ms
  case hours
  case percent
  case breathsPerMin = "breaths_per_min"
  case score
  case event
}

// MARK: - Observation

/// One normalized HealthKit / external sample.
///
/// PHI — never log `value`, `measuredAt`, `sourceRecordId`, or
/// `metadata`. Use `Hashable` for SwiftUI `ForEach` keys without
/// echoing the field values in logs.
struct HealthObservation: Codable, Sendable, Hashable {
  let sourceType: HealthObservationSourceType
  let sourceRecordId: String
  let metricType: HealthObservationMetricType
  let value: Double
  let unit: HealthObservationUnit
  let measuredAt: String
  let startAt: String?
  let endAt: String?

  /// Optional non-PHI structured extensions (e.g.
  /// `["device": "AppleWatch"]`). Carried as a JSON-encoded string so
  /// the struct stays `Hashable` — `[String: Any]` is not Codable and
  /// the contract treats this as an opaque object the backend stores
  /// verbatim. The sync service builds this from a typed dictionary.
  let metadata: HealthObservationMetadata?

  init(
    sourceType: HealthObservationSourceType = .healthkit,
    sourceRecordId: String,
    metricType: HealthObservationMetricType,
    value: Double,
    measuredAt: String,
    startAt: String? = nil,
    endAt: String? = nil,
    metadata: HealthObservationMetadata? = nil,
  ) {
    self.sourceType = sourceType
    self.sourceRecordId = sourceRecordId
    self.metricType = metricType
    self.value = value
    self.unit = metricType.unit
    self.measuredAt = measuredAt
    self.startAt = startAt
    self.endAt = endAt
    self.metadata = metadata
  }
}

/// Typed wrapper around the optional `metadata` object.
///
/// Restricts callers to the small, non-PHI key set the contract
/// allows (`device`, `motionContext`). Adding a free-form `[String:
/// Any]` here would defeat the contract — and would also break
/// `Codable` synthesis. Extend this struct deliberately when a new
/// non-PHI extension is needed.
struct HealthObservationMetadata: Codable, Sendable, Hashable {
  let device: String?
  let motionContext: String?

  init(device: String? = nil, motionContext: String? = nil) {
    self.device = device
    self.motionContext = motionContext
  }

  /// `nil` when no fields are set so the JSON encoder omits the
  /// `metadata` key entirely instead of writing `{}`.
  static func make(device: String? = nil, motionContext: String? = nil)
    -> HealthObservationMetadata?
  {
    if device == nil, motionContext == nil { return nil }
    return HealthObservationMetadata(device: device, motionContext: motionContext)
  }
}

// MARK: - Sync envelope types

/// Request body for `POST /healthkit/sync`.
struct HealthKitSyncRequest: Codable, Sendable {
  let careRecipientId: String
  let observations: [HealthObservation]
}

/// Response body for `POST /healthkit/sync`. Counts only — never
/// echoes per-sample contents back to the client.
struct HealthKitSyncResponse: Codable, Sendable, Hashable {
  let accepted: Int
  let deduped: Int
  let rejected: Int
  let lastSyncedAt: String?
}

/// Response body for `GET /healthkit/status`.
struct HealthKitSyncStatusResponse: Codable, Sendable, Hashable {
  let status: HealthKitSyncStatusValue
  let lastSyncedAt: String?
  let errorMessage: String?
}

/// Discriminated registry status value.
///
/// Mirrors `DataSourceStatus` from `CareRecipientProfile.swift`
/// (intentionally a separate type so a future Phase 4B status drift
/// on the registry side does not silently change the dashboard's
/// view-model contract).
enum HealthKitSyncStatusValue: String, Codable, Sendable, CaseIterable {
  case connected
  case notConnected = "not_connected"
  case error
}

File: shared/models/HealthObservation.js
/**
 * HealthObservation — ESM/JS mirror of the cross-platform sync contract.
 *
 * This file is the backend-consumable mirror of
 * `./HealthObservation.ts`. JSDoc `@typedef` blocks give the backend
 * editor-level type awareness without pulling TypeScript into the Node
 * build. The frozen constant objects below are the runtime enum values —
 * import them from route handlers, validators, and ingest pipelines so
 * every call site agrees on the same strings the TS + Swift contracts use.
 *
 * Rule: if you change a field or value here, you MUST update the `.ts`
 * file, the JSON schema in `shared/contracts/healthObservation.schema.json`,
 * and the Swift model in the same change. Four files, one shape.
 *
 * Phase 4A wires this contract end-to-end:
 *   iOS HealthKitSyncService → POST /healthkit/sync → health_observations
 * Phase 4B will read these rows for baseline + AI summary generation.
 */

// ─── Enum constants (runtime) ───────────────────────────────────────────────

/**
 * Transports that may submit observations. Phase 4A only accepts
 * `healthkit` from the iOS sync companion; `manual` is reserved for
 * caregiver-entered rows in a later phase. Mirrors the
 * `OBSERVATION_SOURCE_TYPES` superset in `apps/backend/lib/
 * health-observations.js` — that constant carries additional values
 * (`apple_health`, `epic`, `healthkit_legacy`) that are not legal for
 * inbound writes from this contract.
 */
export const SYNC_SOURCE_TYPES = Object.freeze({
  healthkit: "healthkit",
  manual: "manual",
})

/**
 * Canonical metric_type strings. Must match `METRIC_TYPES` in
 * `apps/backend/lib/health-observations.js` and the Swift
 * `HealthObservationMetricType` raw values.
 */
export const HEALTH_OBSERVATION_METRIC_TYPES = Object.freeze({
  steps: "steps",
  restingHeartRate: "resting_heart_rate",
  hrv: "hrv",
  spo2: "spo2",
  sleepDuration: "sleep_duration",
  respiratoryRate: "respiratory_rate",
  walkingSteadiness: "walking_steadiness",
  fallEvent: "fall_event",
})

/**
 * Canonical unit strings. Pair 1:1 with metric types so the DB never
 * has to interpret iOS-side enums; see `METRIC_UNIT_BY_METRIC_TYPE`
 * for the allowed pairings.
 */
export const HEALTH_OBSERVATION_UNITS = Object.freeze({
  count: "count",
  bpm: "bpm",
  ms: "ms",
  hours: "hours",
  percent: "percent",
  breathsPerMin: "breaths_per_min",
  score: "score",
  event: "event",
})

/**
 * Allowed unit per metric. Validated server-side at ingest so a
 * contract-broken iOS build cannot insert a `bpm`-typed steps row.
 * The map is the single source of truth — both the JSON schema's
 * `unit` enum and the iOS Codable struct ride this table.
 */
export const HEALTH_OBSERVATION_UNIT_BY_METRIC_TYPE = Object.freeze({
  steps: "count",
  resting_heart_rate: "bpm",
  hrv: "ms",
  spo2: "percent",
  sleep_duration: "hours",
  respiratory_rate: "breaths_per_min",
  walking_steadiness: "score",
  fall_event: "event",
})

// ─── Type aliases for editor tooling (JSDoc) ────────────────────────────────

/**
 * @typedef {"healthkit" | "manual"} HealthObservationSourceType
 *
 * @typedef {"steps" | "resting_heart_rate" | "hrv" | "spo2" | "sleep_duration" | "respiratory_rate" | "walking_steadiness" | "fall_event"} HealthObservationMetricType
 *
 * @typedef {"count" | "bpm" | "ms" | "hours" | "percent" | "breaths_per_min" | "score" | "event"} HealthObservationUnit
 *
 * @typedef {Object} HealthObservation
 * @property {HealthObservationSourceType} sourceType
 * @property {string} sourceRecordId
 * @property {HealthObservationMetricType} metricType
 * @property {number} value
 * @property {HealthObservationUnit} unit
 * @property {string} measuredAt
 * @property {string} [startAt]
 * @property {string} [endAt]
 * @property {Object<string, *>} [metadata]
 *
 * @typedef {Object} HealthKitSyncRequest
 * @property {string} careRecipientId
 * @property {HealthObservation[]} observations
 *
 * @typedef {Object} HealthKitSyncResponse
 * @property {number} accepted
 * @property {number} deduped
 * @property {number} rejected
 * @property {string|null} lastSyncedAt
 *
 * @typedef {Object} HealthKitSyncStatus
 * @property {"connected" | "not_connected" | "error"} status
 * @property {string|null} lastSyncedAt
 * @property {string|null} errorMessage
 */

export {}
